dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/program/notification/ProgramNotificationInstanceServiceTest.java

```diff
@@ -0,0 +1,148 @@
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.security.Authorities;
+import org.hisp.dhis.test.integration.PostgresIntegrationTestBase;
+import org.hisp.dhis.trackedentity.TrackedEntity;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.imports.job.NotificationSender;
+import org.hisp.dhis.tracker.imports.programrule.engine.Notification;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserGroup;
+import org.hisp.dhis.util.DateUtils;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Zubair Asghar
+ */
+@Transactional
+class ProgramNotificationInstanceServiceTest extends PostgresIntegrationTestBase {
+  private static final int TEST_USER_COUNT = 60;
+  private static final Date DATE = DateUtils.parseDate("2025-01-01");
+  private static final int EXPECTED_NOTIFICATIONS = 20;
+
+  @Autowired private ProgramNotificationInstanceService programNotificationInstanceService;
+  @Autowired private IdentifiableObjectManager manager;
+  @Autowired private NotificationSender notificationSender;
+
+  private OrganisationUnit organisationUnit;
+  private UserGroup userGroup;
+  private int expectedNotifications;
+
+  @BeforeEach
+  void setUp() {
+    organisationUnit = createOrganisationUnit('O');
+    manager.save(organisationUnit);
+    Program program = createProgram('P', Set.of(), organisationUnit);
+    manager.save(program);
+    userGroup = createUserGroup('U', new HashSet<>());
+    manager.save(userGroup);
+    createAndAddUsersToUserGroupAndOrgUnit();
+
+    TrackedEntityType trackedEntityType = createTrackedEntityType('P');
+    manager.save(trackedEntityType);
+    TrackedEntity trackedEntity = createTrackedEntity(organisationUnit, trackedEntityType);
+    manager.save(trackedEntity);
+    Enrollment enrollment = createEnrollment(program, trackedEntity, organisationUnit);
+    manager.save(enrollment);
+
+    List<Notification> notifications = getNotifications();
+    expectedNotifications = notifications.size();
+    notifications.forEach(n -> notificationSender.send(n, enrollment));
+  }
+
+  @Test
+  @DisplayName("Should fetch scheduled notifications within timeout")
+  void shouldReturnProgramNotificationInstancesForGivenDate() {
+    ProgramNotificationInstanceParam param =
+        ProgramNotificationInstanceParam.builder().scheduledAt(DATE).build();
+
+    List<ProgramNotificationInstance> instances =
+        programNotificationInstanceService.getProgramNotificationInstances(param);
+
+    assertEquals(
+        expectedNotifications,
+        instances.size(),
+        "Expected " + expectedNotifications + " notification instances, but got " + instances);
+
+    instances.forEach(instance -> assertEquals(DATE, instance.getScheduledAt()));
```

Comment:
Try failing your test by changing the implementation. Would the assertion output help you quickly fix the implementation? Imagine this happens only on ci.